improve deprecated method

saeideng opened the pull request:
This PR is a draft yet

saeideng committed:
fail if no deprecation found

saeideng commented on tests/TestCase/TestSuite/TestCaseTest.php:
```diff
@@ -146,6 +146,32 @@ public function testWithErrorReportingWithException(): void
+        $message = '';
+        try {
+            $this->deprecated(function () {
+            });
```
conflict when passing ```deprecationWarning('test');``` same as previous test `testDeprecated`

but ```deprecationWarning('other string');``` will stop assert failing in ` $this->assertTrue($deprecation, ...`

saeideng replied:
this was new feature
Refs: https://github.com/cakephp/cakephp/commit/eef535c1153a9

markstory commented on src/TestSuite/TestCase.php:
```diff
@@ -203,13 +203,27 @@ public function withErrorReporting(int $errorLevel, callable $callable): void
     {
-        $errorLevel = error_reporting();
-        error_reporting($errorLevel & ~E_USER_DEPRECATED);
+        $deprecation = false;
```
```suggestion
        /** @var boolean $deprecation */
        $deprecation = false;
```

This might help with the error on 226

markstory commented on tests/TestCase/TestSuite/TestCaseTest.php:
```diff
@@ -146,6 +146,32 @@ public function testWithErrorReportingWithException(): void
+        } catch (\Exception $e) {
+            $message = $e->getMessage();
+        } finally {
+            $this->assertStringStartsWith('Should have at least one deprecation warning', $message);
```
You could have this assertion in the catch block and skip finally. You can also use `$this->fail()` at the bottom of the try to ensure an exception was raised.

saeideng replied:
done

saeideng committed:
Update src/TestSuite/TestCase.php

Co-authored-by: Mark Story <mark@mark-story.com>

saeideng committed:
change code style / test deprecated() method

since duplicated deprecations will remove
I used the trigger_error

Refs : https://github.com/cakephp/cakephp/commit/eef535c1153

othercorey commented:
All the tests are still failing.

saeideng commented:
> All the tests are still failing.

yes now shows , some methods don't  have deprecation warning

```

1) Cake\Test\TestCase\Controller\Component\PaginatorComponentTest::testDeprecatedPathMergeOptionsExtraWhitelist
Failed asserting that exception of type "PHPUnit\Framework\Error\Deprecated" is thrown.

2) Cake\Test\TestCase\Controller\ComponentTest::testDeprecatedEventShutdownPath
Failed asserting that exception of type "PHPUnit\Framework\Error\Deprecated" is thrown.
```
these are strange


addressed failures have not `deprecated()` method , but other test case above it  have, with error ... 

and when test these methods  individually no error occurs
```
vendor/bin/phpunit tests/TestCase --filter testDeprecatedEventShutdownPath
vendor/bin/phpunit tests/TestCase --filter testDeprecatedPathMergeOptionsExtraWhitelist
```

I think these wrong errors will fix , when we fix correct methods/testCases

markstory commented:
> I think these wrong errors will fix , when we fix correct methods/testCases

We could remove the usage of `deprecated()` in those tests as they fail when run in bulk.

othercorey commented:
> We could remove the usage of `deprecated()` in those tests as they fail when run in bulk.

I think that is hiding some other issue. My guess is the deprecated flag is removed by some other test.

saeideng committed:
test deprecated with assert after trigger warning

saeideng commented on tests/TestCase/TestSuite/TestCaseTest.php:
```diff
@@ -146,6 +146,48 @@ public function testWithErrorReportingWithException(): void
+        try {
+            $this->deprecated(function () {
+                trigger_error('deprecation message', E_USER_DEPRECATED);
+                $this->assertTrue(false, 'A random message');
```
this testCase tries to show `deprecated()` behave unlike the phpunit's `$this->expectDeprecation();`

saeideng committed:
fix not throwing deprecation warning

1)removing `testAllowEmptyDateTimeDeprecated`:

deprecated argument order was in cake 3.x , not cake4
also this testCase was deplicated with
`testAllowEmptyDateTimeCondition`

2) throwing deprecation warning in testArrayPropertyMerge and
testArrayProperties

second Argument of setHelpers deprecated in cake +4.3
and in cake5 should use addHelper()/addHelpers() instead.

saeideng commented:
> I think that is hiding some other issue. My guess is the deprecated flag is removed by some other test.

it was because of these lines
https://github.com/cakephp/cakephp/blob/219de3bab07ff983662b9f75d9d32fdafaa6d9aa/src/Core/functions.php#L312-L317

othercorey commented on tests/TestCase/Validation/ValidatorTest.php:
```diff
@@ -1300,28 +1300,6 @@ public function testAllowEmptyDateTimeCondition(): void
-        $validator = new Validator();
-        $this->deprecated(function () use ($validator): void {
-            $validator->allowEmptyDateTime('published', 'datetime required', 'update');
-        });
```
Is this allowed now?

saeideng replied:
see above testCase , it only is duplicated TestCase but without `deprecated()` right now
not sure what was this testCase before

`$when` as second Argument allowed (with deprecation) in cake 3 (not sure after which version)
but now in cake 4 `$when` as second Argument is not allowed anymore

markstory commented on tests/TestCase/Core/FunctionsTest.php:
```diff
@@ -91,6 +91,30 @@ public function testDeprecationWarningEnabled(): void
     }
 
+    /**
+     * Test duplicate deprecations emit in cli environment and when run PHPUnit
```
Won't emitting duplicates have the drawback of having more noisy output in tests? My thinking was that you might have a method that is called 50+ times and that you don't need a deprecation warning for each time it is called.

saeideng replied:
when we have several testCases each those need to have /catch deprecation warning independently
I had not good Idea to both , catching deprecation with `deprecated()` + not annoying developer

if there is better solution I can revert last commit

saeideng replied:
@markstory  there is a way

having other condition in `deprecationWarning()` method
finding deprecated() method in `$trace` and showing all deprecations, otherwise ignoring duplicated warning

I think it is available
if you agree I can do it in next PR

markstory replied:
> finding deprecated() method in $trace and showing all deprecations, otherwise ignoring duplicated warning

I think that is a reasonable solution. Would we only crawl the stack if the message was identified as a duplicate? Walking the stack could be expensive and create measurable overhead in application code at the beginning of their upgrade process.

saeideng replied:
> Would we only crawl the stack if the message was identified as a duplicate?

yes , right now in this PR is like the

```
        if (
                isset($errors[$checksum]) &&
                (
                    // show all deprecations when phpunit runs
                    //conditions
                )
        )
```

markstory commented on tests/TestCase/TestSuite/TestCaseTest.php:
```diff
@@ -146,6 +146,48 @@ public function testWithErrorReportingWithException(): void
+    public function testDeprecated(): void
+    {
+        $this->deprecated(function () {
+            trigger_error('deprecation message', E_USER_DEPRECATED);
```
You could make different messages to get around the duplicate prevention 🤷

saeideng replied:
this commit was before my last commit

saeideng replied:
and we just need a deprecation warning not complex `deprecationwarning()`, and this helps us to maintaining easier
if needed I can put a  `deprecationwarning()`

markstory replied:
I made this suggestion when I thought the removal of the duplicate detection was because it made writing these tests easier. However, that was before I knew about the additional use-cases you outlined in https://github.com/cakephp/cakephp/pull/15935#discussion_r716800070. I no longer think anything needs to change from the current diff.

saeideng commented:
about remains failures
```
    public function test1(): void
    {
        $this->deprecated(function () {
            $obj->method();
        });
    }

    public function test2(): void
    {
           $this->expectDeprecation();
           $this->expectDeprecationMessage('....');
           $obj->method();
    }
```

now we story checksum `$errors[$checksum] = true;` in first testCase `test1`
both use same method ` $obj->method();` so there are duplicated message 
and `test2` expect to see deprecation


because we only allow to triggering  duplicated warning when using the `deprecated`
and `test2` do not use it , it fails 
```
Failed asserting that exception of type "PHPUnit\Framework\Error\Deprecated" is thrown.
```

markstory commented on src/Core/functions.php:
```diff
@@ -308,15 +310,35 @@ function deprecationWarning(string $message, int $stackFrame = 1): void
+            }
+
+            return false;
+        };
```
This will work fine, but should we be doing something 'dumber' and use a `Configure` value, or global variable? `TestCase::deprecated()` could set the global state and `deprecationWarning()` could check if it's set. Using a configure variable of giving users an easy way to have all their deprecations display if they wanted that.

saeideng replied:
OK, 
I can remove some current code,
and then adding a static variable to `TestCase` , maybe helps

```
TestCase::showDuplicatedDeprecation = false
```
we can access it from `deprecated()` and `deprecationWarning()`

we can add enabler method in future , depended to feedback
user only needs to enable it when they want to check with `expectDeprecation` , when there are 2 times

also we can keep the `$errors['tests']` to not check the `showDuplicatedDeprecation` in production
let me to try it

saeideng replied:
@markstory I just added global config, and removed the complexity

>Using a configure variable of giving users an easy way to have all their deprecations display if they wanted that.

yes I added `Configure` , user could enable it to show all duplicated deprecations

saeideng replied:
I did not add `TestCase::showDuplicatedDeprecation` because all people not use cake's testSuite
or if someone want to use cake's repo(s) standalone , or even if they want to show all in web (debug time ) 
having global setting is more useful

saeideng committed:
add Error.enableDuplicatedDeprecations

enable duplicated warning to catch by TestSuite::deprecated()
method

saeideng commented on src/Core/functions.php:
```diff
@@ -311,10 +311,13 @@ function deprecationWarning(string $message, int $stackFrame = 1): void
-        $errors[$checksum] = true;
+        if (!$duplicate) {
+            $errors[$checksum] = true;
+        }
```
this causes to not keep `checksum` if user want to show all deprecations

so if you use `TestSuite::deprecated()` this prevent to save the `checksum` , this allows us to catch all deprecations by `TestSuite::deprecated()`

saeideng replied:
also this preventing helps to use `$this->expectDeprecation()` after using `TestSuite::deprecated()`

but if user had 2 duplicate deprecation + `$this->expectDeprecation()`

he should enable duplicated deprecation

markstory commented on src/TestSuite/TestCase.php:
```diff
@@ -203,13 +203,37 @@ public function withErrorReporting(int $errorLevel, callable $callable): void
     {
-        $errorLevel = error_reporting();
-        error_reporting($errorLevel & ~E_USER_DEPRECATED);
+        $duplicate = Configure::read('Error.enableDuplicatedDeprecations');
```
You shouldn't need to make this variable or restore the value at the end of the test method. The `tearDown()` of `TestCase` will restore all of `Configure` anyways.

saeideng replied:
not in this case
1)
```php

    public function testMethod(): void
{
           $this->expectDeprecation();
           $this->expectDeprecationMessage('after deprecated')
           $this->deprecated(function () {
                     trigger_error('message within deprecated method', E_USER_DEPRECATED);
                     trigger_error('other message', E_USER_DEPRECATED);
            });

            trigger_error('after deprecated', E_USER_DEPRECATED);
}
```
we want to above testCase passes

```php
    public function testMethod(): void
{
           // section 1
           $this->deprecated(function () {
                     // section 2
            });

            // section 3
}
```

by default in `section 1` and `section 3` deprecations will trigger unless if there was duplicate deprecation
(not included warnings from `deprecated()` 

in `section 2` we must trigger all deprecation even duplicated, ( but we should not include them into `checksum` )

`deprecated()`  just works like before this PR,
we dont want to change other sections 

so we have to back to previous Config immediately after finishing the `deprecated()` block


2) by  default we prevent to show duplicated deprecation unless user want to use `deprecated()` 
so we did not change this Config before (in `setUp` method ), so we can not understand what was before to restore it in `tearDown()`

markstory replied:
I did not consider that scenario. Thank you for the great explanation :bow:

markstory reviewed (approved):
Looks good to me.

markstory reviewed (approved):
Looks good to me.
